File: backend/src/main/java/com/bitacora/application/notification/CollaborationService.java
package com.bitacora.application.notification;

import com.bitacora.domain.model.activity.Activity;
import com.bitacora.domain.model.notification.CollaborationNotification;
import com.bitacora.domain.model.notification.CollaborationNotification.CollaborationAction;
import com.bitacora.domain.model.user.User;
import com.bitacora.domain.port.notification.NotificationPort;
import com.bitacora.domain.port.repository.ActivityRepository;
import com.bitacora.domain.port.UserRepository;

import org.springframework.stereotype.Service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.Set;
import java.util.concurrent.ConcurrentSkipListSet;

/**
 * Servicio para manejar la colaboración en tiempo real.
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class CollaborationService {

    private final NotificationPort notificationPort;
    private final ActivityRepository activityRepository;
    private final UserRepository userRepository;

    // Mapa para rastrear qué usuarios están viendo/editando cada actividad
    private final ConcurrentMap<Long, Set<Long>> activityViewers = new ConcurrentHashMap<>();
    private final ConcurrentMap<Long, Long> activityEditors = new ConcurrentHashMap<>();

    /**
     * Registra que un usuario está viendo una actividad.
     *
     * @param activityId El ID de la actividad
     * @param userId     El ID del usuario
     * @return true si se registró correctamente, false en caso contrario
     */
    public boolean registerViewer(Long activityId, Long userId) {
        log.debug("Registrando usuario {} como visor de la actividad {}", userId, activityId);

        // Obtener la actividad
        Optional<Activity> activityOpt = activityRepository.findById(activityId);
        if (activityOpt.isEmpty()) {
            log.warn("No se encontró la actividad con ID: {}", activityId);
            return false;
        }

        // Obtener el usuario
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            log.warn("No se encontró el usuario con ID: {}", userId);
            return false;
        }

        Activity activity = activityOpt.get();
        User user = userOpt.get();

        // Registrar al usuario como visor
        Set<Long> viewers = activityViewers.computeIfAbsent(activityId, k -> new ConcurrentSkipListSet<>());
        boolean isNewViewer = viewers.add(userId);

        // Si es un nuevo visor, notificar a los demás usuarios
        if (isNewViewer) {
            CollaborationNotification notification = CollaborationNotification.builder()
                    .activityId(activityId)
                    .activityTitle(activity.getDescription())
                    .userId(userId)
                    .userName(user.getPersonName().getFullName())
                    .action(CollaborationAction.VIEWING)
                    .title("Usuario viendo actividad")
                    .message(user.getPersonName().getFullName() + " está viendo la actividad")
                    .build();

            notificationPort.broadcastCollaborationNotification(notification, activityId);
        }

        return true;
    }

    /**
     * Registra que un usuario está editando una actividad.
     *
     * @param activityId El ID de la actividad
     * @param userId     El ID del usuario
     * @return true si se registró correctamente, false en caso contrario
     */
    public boolean registerEditor(Long activityId, Long userId) {
        log.debug("Registrando usuario {} como editor de la actividad {}", userId, activityId);

        // Obtener la actividad
        Optional<Activity> activityOpt = activityRepository.findById(activityId);
        if (activityOpt.isEmpty()) {
            log.warn("No se encontró la actividad con ID: {}", activityId);
            return false;
        }

        // Obtener el usuario
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            log.warn("No se encontró el usuario con ID: {}", userId);
            return false;
        }

        Activity activity = activityOpt.get();
        User user = userOpt.get();

        // Verificar si ya hay un editor
        Long currentEditor = activityEditors.get(activityId);
        if (currentEditor != null && !currentEditor.equals(userId)) {
            log.warn("La actividad {} ya está siendo editada por el usuario {}", activityId, currentEditor);
            return false;
        }

        // Registrar al usuario como editor
        activityEditors.put(activityId, userId);

        // Registrar también como visor
        registerViewer(activityId, userId);

        // Notificar a los demás usuarios
        CollaborationNotification notification = CollaborationNotification.builder()
                .activityId(activityId)
                .activityTitle(activity.getDescription())
                .userId(userId)
                .userName(user.getPersonName().getFullName())
                .action(CollaborationAction.EDITING)
                .title("Usuario editando actividad")
                .message(user.getPersonName().getFullName() + " está editando la actividad")
                .build();

        notificationPort.broadcastCollaborationNotification(notification, activityId);

        return true;
    }

    /**
     * Registra que un usuario ha comentado en una actividad.
     *
     * @param activityId El ID de la actividad
     * @param userId     El ID del usuario
     * @param comment    El comentario
     * @return true si se registró correctamente, false en caso contrario
     */
    public boolean registerComment(Long activityId, Long userId, String comment) {
        log.debug("Registrando comentario del usuario {} en la actividad {}", userId, activityId);

        // Obtener la actividad
        Optional<Activity> activityOpt = activityRepository.findById(activityId);
        if (activityOpt.isEmpty()) {
            log.warn("No se encontró la actividad con ID: {}", activityId);
            return false;
        }

        // Obtener el usuario
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            log.warn("No se encontró el usuario con ID: {}", userId);
            return false;
        }

        Activity activity = activityOpt.get();
        User user = userOpt.get();

        // Notificar a los demás usuarios
        CollaborationNotification notification = CollaborationNotification.builder()
                .activityId(activityId)
                .activityTitle(activity.getDescription())
                .userId(userId)
                .userName(user.getPersonName().getFullName())
                .action(CollaborationAction.COMMENTED)
                .title("Nuevo comentario en actividad")
                .message(user.getPersonName().getFullName() + " ha comentado: " + comment)
                .build();

        notificationPort.broadcastCollaborationNotification(notification, activityId);

        return true;
    }

    /**
     * Registra que un usuario ha dejado de ver/editar una actividad.
     *
     * @param activityId El ID de la actividad
     * @param userId     El ID del usuario
     * @return true si se registró correctamente, false en caso contrario
     */
    public boolean unregisterUser(Long activityId, Long userId) {
        log.debug("Eliminando usuario {} como visor/editor de la actividad {}", userId, activityId);

        // Obtener la actividad
        Optional<Activity> activityOpt = activityRepository.findById(activityId);
        if (activityOpt.isEmpty()) {
            log.warn("No se encontró la actividad con ID: {}", activityId);
            return false;
        }

        // Obtener el usuario
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            log.warn("No se encontró el usuario con ID: {}", userId);
            return false;
        }

        Activity activity = activityOpt.get();
        User user = userOpt.get();

        // Eliminar al usuario como editor si lo es
        Long currentEditor = activityEditors.get(activityId);
        if (currentEditor != null && currentEditor.equals(userId)) {
            activityEditors.remove(activityId);
        }

        // Eliminar al usuario como visor
        Set<Long> viewers = activityViewers.get(activityId);
        if (viewers != null) {
            boolean wasViewer = viewers.remove(userId);

            // Si estaba viendo la actividad, notificar a los demás usuarios
            if (wasViewer) {
                CollaborationNotification notification = CollaborationNotification.builder()
                        .activityId(activityId)
                        .activityTitle(activity.getDescription())
                        .userId(userId)
                        .userName(user.getPersonName().getFullName())
                        .action(CollaborationAction.LEFT)
                        .title("Usuario dejó de ver actividad")
                        .message(user.getPersonName().getFullName() + " dejó de ver la actividad")
                        .build();

                notificationPort.broadcastCollaborationNotification(notification, activityId);
            }

            // Si no hay más visores, eliminar la entrada del mapa
            if (viewers.isEmpty()) {
                activityViewers.remove(activityId);
            }
        }

        return true;
    }

    /**
     * Obtiene el conjunto de usuarios que están viendo una actividad.
     *
     * @param activityId El ID de la actividad
     * @return El conjunto de IDs de usuarios
     */
    public Set<Long> getViewers(Long activityId) {
        return activityViewers.getOrDefault(activityId, new ConcurrentSkipListSet<>());
    }

    /**
     * Obtiene el usuario que está editando una actividad.
     *
     * @param activityId El ID de la actividad
     * @return El ID del usuario, o null si nadie está editando la actividad
     */
    public Long getEditor(Long activityId) {
        return activityEditors.get(activityId);
    }
}

File: frontend/src/services/collaborationService.ts
import { websocketService } from './websocketService';

/**
 * Servicio para manejar la colaboración en tiempo real.
 */
class CollaborationService {
  private activityViewers: Map<number, Set<number>> = new Map();
  private activityEditors: Map<number, number> = new Map();
  private currentUser: number | null = null;
  private isInitialized: boolean = false;
  private connectedUsers: Map<number, { userName: string, lastActivity: number }> = new Map();
  private userActivityTimeout: number = 5 * 60 * 1000; // 5 minutos de inactividad antes de considerar desconectado

  /**
   * Inicializa el servicio de colaboración.
   *
   * @param userId El ID del usuario actual
   */
  public init(userId: number): void {
    if (this.isInitialized) {
      return;
    }

    this.currentUser = userId;
    this.isInitialized = true;
    console.log('Servicio de colaboración inicializado para el usuario:', userId);
  }

  /**
   * Registra que el usuario actual está viendo una actividad.
   *
   * @param activityId El ID de la actividad
   */
  public registerViewer(activityId: number): void {
    if (!this.isInitialized || !this.currentUser) {
      console.warn('El servicio de colaboración no está inicializado');
      return;
    }

    console.log(`Registrando usuario ${this.currentUser} como visor de la actividad ${activityId}`);

    // Enviar evento al servidor
    websocketService.send('collaboration_view', {
      activityId,
      userId: this.currentUser
    });

    // Actualizar estado local
    if (!this.activityViewers.has(activityId)) {
      this.activityViewers.set(activityId, new Set());
    }
    this.activityViewers.get(activityId)?.add(this.currentUser);
  }

  /**
   * Registra que el usuario actual está editando una actividad.
   *
   * @param activityId El ID de la actividad
   */
  public registerEditor(activityId: number): void {
    if (!this.isInitialized || !this.currentUser) {
      console.warn('El servicio de colaboración no está inicializado');
      return;
    }

    console.log(`Registrando usuario ${this.currentUser} como editor de la actividad ${activityId}`);

    // Enviar evento al servidor
    websocketService.send('collaboration_edit', {
      activityId,
      userId: this.currentUser
    });

    // Actualizar estado local
    this.activityEditors.set(activityId, this.currentUser);
    this.registerViewer(activityId); // También registrar como visor
  }

  /**
   * Registra que el usuario actual ha comentado en una actividad.
   *
   * @param activityId El ID de la actividad
   * @param comment El comentario
   */
  public registerComment(activityId: number, comment: string): void {
    if (!this.isInitialized || !this.currentUser) {
      console.warn('El servicio de colaboración no está inicializado');
      return;
    }

    console.log(`Registrando comentario del usuario ${this.currentUser} en la actividad ${activityId}`);

    // Enviar evento al servidor
    websocketService.send('collaboration_comment', {
      activityId,
      userId: this.currentUser,
      comment
    });
  }

  /**
   * Registra que el usuario actual ha dejado de ver/editar una actividad.
   *
   * @param activityId El ID de la actividad
   */
  public unregisterUser(activityId: number): void {
    if (!this.isInitialized || !this.currentUser) {
      console.warn('El servicio de colaboración no está inicializado');
      return;
    }

    console.log(`Eliminando usuario ${this.currentUser} como visor/editor de la actividad ${activityId}`);

    // Enviar evento al servidor
    websocketService.send('collaboration_leave', {
      activityId,
      userId: this.currentUser
    });

    // Actualizar estado local
    const viewers = this.activityViewers.get(activityId);
    if (viewers) {
      viewers.delete(this.currentUser);
      if (viewers.size === 0) {
        this.activityViewers.delete(activityId);
      }
    }

    if (this.activityEditors.get(activityId) === this.currentUser) {
      this.activityEditors.delete(activityId);
    }
  }

  /**
   * Actualiza la lista de usuarios que están viendo una actividad.
   *
   * @param activityId El ID de la actividad
   * @param userIds Los IDs de los usuarios
   */
  public updateViewers(activityId: number, userIds: number[]): void {
    const viewers = new Set(userIds);
    this.activityViewers.set(activityId, viewers);
  }

  /**
   * Actualiza el usuario que está editando una actividad.
   *
   * @param activityId El ID de la actividad
   * @param userId El ID del usuario
   */
  public updateEditor(activityId: number, userId: number | null): void {
    if (userId) {
      this.activityEditors.set(activityId, userId);
    } else {
      this.activityEditors.delete(activityId);
    }
  }

  /**
   * Obtiene los usuarios que están viendo una actividad.
   *
   * @param activityId El ID de la actividad
   * @returns Los IDs de los usuarios
   */
  public getViewers(activityId: number): number[] {
    return Array.from(this.activityViewers.get(activityId) || []);
  }

  /**
   * Obtiene el usuario que está editando una actividad.
   *
   * @param activityId El ID de la actividad
   * @returns El ID del usuario, o null si nadie está editando la actividad
   */
  public getEditor(activityId: number): number | null {
    return this.activityEditors.get(activityId) || null;
  }

  /**
   * Verifica si el usuario actual está viendo una actividad.
   *
   * @param activityId El ID de la actividad
   * @returns true si el usuario está viendo la actividad, false en caso contrario
   */
  public isViewing(activityId: number): boolean {
    if (!this.currentUser) return false;
    return this.activityViewers.get(activityId)?.has(this.currentUser) || false;
  }

  /**
   * Verifica si el usuario actual está editando una actividad.
   *
   * @param activityId El ID de la actividad
   * @returns true si el usuario está editando la actividad, false en caso contrario
   */
  public isEditing(activityId: number): boolean {
    if (!this.currentUser) return false;
    return this.activityEditors.get(activityId) === this.currentUser;
  }

  /**
   * Verifica si alguien más está editando una actividad.
   *
   * @param activityId El ID de la actividad
   * @returns true si alguien más está editando la actividad, false en caso contrario
   */
  public isSomeoneElseEditing(activityId: number): boolean {
    const editor = this.activityEditors.get(activityId);
    return !!editor && editor !== this.currentUser;
  }

  /**
   * Obtiene el número de usuarios que están viendo una actividad.
   *
   * @param activityId El ID de la actividad
   * @returns El número de usuarios
   */
  public getViewerCount(activityId: number): number {
    return this.activityViewers.get(activityId)?.size || 0;
  }

  /**
   * Limpia el estado del servicio.
   */
  public clear(): void {
    this.activityViewers.clear();
    this.activityEditors.clear();
    this.currentUser = null;
    this.isInitialized = false;
    this.connectedUsers.clear();
  }

  /**
   * Registra un usuario como conectado.
   *
   * @param userId El ID del usuario
   * @param userName El nombre del usuario
   */
  public registerConnectedUser(userId: number, userName: string): void {
    this.connectedUsers.set(userId, {
      userName,
      lastActivity: Date.now()
    });

    // Enviar evento al servidor
    websocketService.send('user_connected', {
      userId,
      userName,
      connectionTime: Date.now()
    });

    console.log(`Usuario ${userName} (${userId}) registrado como conectado`);
  }

  /**
   * Actualiza la actividad de un usuario.
   *
   * @param userId El ID del usuario
   */
  public updateUserActivity(userId: number): void {
    const user = this.connectedUsers.get(userId);
    if (user) {
      this.connectedUsers.set(userId, {
        ...user,
        lastActivity: Date.now()
      });
    }
  }

  /**
   * Registra un usuario como desconectado.
   *
   * @param userId El ID del usuario
   */
  public registerDisconnectedUser(userId: number): void {
    const user = this.connectedUsers.get(userId);
    if (user) {
      // Enviar evento al servidor
      websocketService.send('user_disconnected', {
        userId,
        userName: user.userName,
        disconnectionTime: Date.now(),
        sessionDuration: Math.floor((Date.now() - user.lastActivity) / 1000)
      });

      console.log(`Usuario ${user.userName} (${userId}) registrado como desconectado`);
      this.connectedUsers.delete(userId);
    }
  }

  /**
   * Obtiene la lista de usuarios conectados.
   *
   * @returns Lista de usuarios conectados
   */
  public getConnectedUsers(): { userId: number, userName: string, lastActivity: number }[] {
    const now = Date.now();
    const activeUsers: { userId: number, userName: string, lastActivity: number }[] = [];

    // Filtrar usuarios inactivos
    this.connectedUsers.forEach((user, userId) => {
      if (now - user.lastActivity < this.userActivityTimeout) {
        activeUsers.push({
          userId,
          userName: user.userName,
          lastActivity: user.lastActivity
        });
      } else {
        // Registrar como desconectado si ha pasado el tiempo de inactividad
        this.registerDisconnectedUser(userId);
      }
    });

    return activeUsers;
  }

  /**
   * Verifica si un usuario está conectado.
   *
   * @param userId El ID del usuario
   * @returns true si el usuario está conectado
   */
  public isUserConnected(userId: number): boolean {
    const user = this.connectedUsers.get(userId);
    if (!user) return false;

    const now = Date.now();
    if (now - user.lastActivity >= this.userActivityTimeout) {
      // Registrar como desconectado si ha pasado el tiempo de inactividad
      this.registerDisconnectedUser(userId);
      return false;
    }

    return true;
  }
}

// Exportar una instancia única del servicio
export const collaborationService = new CollaborationService();

/**
 * Hook para usar el servicio de colaboración.
 */
export const useCollaboration = () => {
  return {
    init: collaborationService.init.bind(collaborationService),
    registerViewer: collaborationService.registerViewer.bind(collaborationService),
    registerEditor: collaborationService.registerEditor.bind(collaborationService),
    registerComment: collaborationService.registerComment.bind(collaborationService),
    unregisterUser: collaborationService.unregisterUser.bind(collaborationService),
    getViewers: collaborationService.getViewers.bind(collaborationService),
    getEditor: collaborationService.getEditor.bind(collaborationService),
    isViewing: collaborationService.isViewing.bind(collaborationService),
    isEditing: collaborationService.isEditing.bind(collaborationService),
    isSomeoneElseEditing: collaborationService.isSomeoneElseEditing.bind(collaborationService),
    getViewerCount: collaborationService.getViewerCount.bind(collaborationService),
    updateViewers: collaborationService.updateViewers.bind(collaborationService),
    updateEditor: collaborationService.updateEditor.bind(collaborationService),
    clear: collaborationService.clear.bind(collaborationService),
    // Nuevos métodos para gestionar usuarios conectados
    registerConnectedUser: collaborationService.registerConnectedUser.bind(collaborationService),
    updateUserActivity: collaborationService.updateUserActivity.bind(collaborationService),
    registerDisconnectedUser: collaborationService.registerDisconnectedUser.bind(collaborationService),
    getConnectedUsers: collaborationService.getConnectedUsers.bind(collaborationService),
    isUserConnected: collaborationService.isUserConnected.bind(collaborationService)
  };
};
